lib/source-code/source-code.js

```diff
@@ -681,6 +681,36 @@ class SourceCode extends TokenStore {
     }
     /* eslint-enable class-methods-use-this -- node is owned by SourceCode */
 
+    /**
+     * Marks a variable as used in the current scope
+     * @param {string} name The name of the variable to mark as used.
+     * @param {ASTNode} [refNode] The closest node to the variable reference.
+     * @returns {boolean} True if the variable was found and marked as used, false if not.
+     */
+    markVariableAsUsed(name, refNode = this.ast) {
+
+        const currentScope = this.getScope(refNode);
+        const hasSpecialScope = this.scopeManager.isGlobalReturn() ||
+            this.scopeManager.isModule();
```

Comment:
This looks like a breaking change because custom scope managers are not required to provide these methods.

In the [`parseForESLint`](https://eslint.org/docs/latest/extend/custom-parsers#parseforeslint-return-object) interface specification, we require parsers to return a [`ScopeManager`](https://eslint.org/docs/latest/extend/scope-manager-interface) object whose interface doesn't mention `isGlobalReturn()` and lists `isModule()` as deprecated and not used in ESLint:

https://github.com/eslint/eslint/blob/f57eff20f5789408e95061f1af5354bb9b4f4784/docs/src/extend/scope-manager-interface.md?plain=1#L43-L51

Reply:
I think the goal here is just, when the Program node is passed, to return the innermost top-level scope instead of the outermost (global) scope that `getScope()` returns, so if we want to avoid analyzing configured language options, perhaps we could check the scopes directly:

```js
const initialScope = currentScope.type === "global" && currentScope.childScopes.length > 0 && currentScope.childScopes[0].block === this.ast
            ? currentScope.childScopes[0]
            : currentScope;
```

or maybe do something like this:

```js
const initialScope = refNode === this.ast
            ? this.scopeManager.acquire(refNode, /* inner = */ true)
            : this.getScope(refNode);
```